File: 0125-valid-palindrome/0125-valid-palindrome.js
/**
 * @param {string} s
 * @return {boolean}
 */

const validChars = new Set("abcdefghijklmnopqrstuvwxyz0123456789");
var isPalindrome = function(s) {
    const reduced = [...s.toLowerCase()].filter(c => validChars.has(c)).join('');
    let left = 0;
    let right = reduced.length-1;
    while (left < right) {
        if (reduced[left++] != reduced[right--]) return false;
    }
    return true;
};

File: 0125-valid-palindrome/0125-valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        left = 0
        right = len(s)-1
        s = s.lower()

        while left < right:
            while not s[left].isalnum():
                left+=1
                if left > right: return True
            while not s[right].isalnum():
                right-=1
                if left > right: return True
            
            if s[left] == s[right]:
                left+=1
                right-=1
            else:
                return False
        
        return True